pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java

```diff
@@ -387,4 +418,182 @@ private long calculateMemoryForCompletedSegmentsPerPartition(long completedSegme
+      });
+      schema.getAllFieldSpecs().forEach(fieldSpec -> {
+        String name = fieldSpec.getName();
+        dataTypes.put(name, fieldSpec.getDataType());
+        fieldTypes.put(name, fieldSpec.getFieldType());
+      });
+      timeUnits.put(schema.getTimeFieldSpec().getName(),
+          schema.getTimeFieldSpec().getIncomingGranularitySpec().getTimeType());
+
+      // generate data
+      String outputDir = getOutputDir(now, "-csv");
+      DataGeneratorSpec spec =
+          new DataGeneratorSpec(colNames, cardinalities, new HashMap<>(), new HashMap<>(), mvCounts, lengths, dataTypes,
+              fieldTypes, timeUnits, FileFormat.CSV, outputDir, true);
+      DataGenerator dataGenerator = new DataGenerator();
+      try {
+        dataGenerator.init(spec);
+        dataGenerator.generateCsv(dataCharacteristics.numberOfRows, 1);
+        File outputFile = Paths.get(outputDir, "output_0.csv").toFile();
+        LOGGER.info("Successfully generated data file: {}", outputFile);
+        return outputFile;
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private File createSegment(File csvDataFile, Schema schema, Date now) {
+
+      // create segment
+      LOGGER.info("Started creating segment from file: {}", csvDataFile);
+      String outDir = getOutputDir(now, "-segment");
+      SegmentGeneratorConfig segmentGeneratorConfig = getSegmentGeneratorConfig(csvDataFile, schema, outDir);
+      SegmentIndexCreationDriver driver = new SegmentIndexCreationDriverImpl();
+      try {
+        driver.init(segmentGeneratorConfig);
+        driver.build();
+      } catch (Exception e) {
+        throw new RuntimeException("Caught exception while generating segment from file: " + csvDataFile, e);
+      }
+      String segmentName = driver.getSegmentName();
+      File indexDir = new File(outDir, segmentName);
+      LOGGER.info("Successfully created segment: {} at directory: {}", segmentName, indexDir);
+
+      // verify segment
+      LOGGER.info("Verifying the segment by loading it");
+      ImmutableSegment segment;
+      try {
+        segment = ImmutableSegmentLoader.load(indexDir, ReadMode.mmap);
+      } catch (Exception e) {
+        throw new RuntimeException("Caught exception while verifying the created segment", e);
+      }
+      LOGGER.info("Successfully loaded segment: {} of size: {} bytes", segmentName,
+          segment.getSegmentSizeBytes());
+      segment.destroy();
+
+      return indexDir;
+    }
+
+    private SegmentGeneratorConfig getSegmentGeneratorConfig(File csvDataFile, Schema schema, String outDir) {
+      SegmentGeneratorConfig segmentGeneratorConfig = new SegmentGeneratorConfig(_tableConfig, schema);
+      segmentGeneratorConfig.setInputFilePath(csvDataFile.getPath());
+      segmentGeneratorConfig.setFormat(FileFormat.CSV);
+      segmentGeneratorConfig.setOutDir(outDir);
+      segmentGeneratorConfig.setReaderConfig(new CSVRecordReaderConfig()); // FIXME
```

Comment:
Why FIXME?

Reply:
It was for my book-keeping that got away to get cleaned up before sending the PR. It's removed now.